### PR TITLE
[MLIR][XeGPU] Adding XeGPU 2d block operators

### PR DESCRIPTION
**Pull request by chencha3**
Hi @joker-eph, This PR adds XeGPU 2D block operators. It contains:
1. `TensorDescType` and `TensorDescAttr` definitions
2. `MemoryScopeAttr` and `CacheHintAttr` definitions which are used by `TensorDescAttr`. 
3. `CreateNdDescOp`, `PrefetchNdOp`, `LoadNdOp`, and `StoreNdOp` definitions, and their corresponding testcases for illustration.

### CODE REVIEW

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 1]
```diff
@@ -23,4 +39,199 @@ class XeGPU_Op<string mnemonic, list<Trait> traits = []>:
           Op<XeGPU_Dialect, mnemonic, traits>;
 
 
+def XeGPU_CreateNdDescOp: XeGPU_Op<"create_nd_tdesc", [Pure, AttrSizedOperandSegments]> {
+
+  let summary = "create nd tensor descriptor operation";
+  let description = [{
+    The "create_nd_tdesc" operation creates a TensorDescType which represents
+    a sub-view of a 2D memory region (It can be extended to support N-D memory
```
Comment (adam-smnk) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:47]: As per `XeGPU_BaseAddrType`, 1D memref is also accepted and verifier also doesn't complain.
Which behavior is intended?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 2]
```diff
@@ -23,4 +39,199 @@ class XeGPU_Op<string mnemonic, list<Trait> traits = []>:
...
+
+    /// Return the shape of the TensorDesc
+    llvm::ArrayRef<int64_t> getTensorDescShape() {
+      return getType().getShape();
+    }
+  }];
+}
+
+def XeGPU_PrefetchNdOp : XeGPU_Op<"prefetch_nd", []> {
+  let summary = "prefetches a nD block to cache";
```
Comment (adam-smnk) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:167]: nit: unify "nD/n-D/N-D block" style

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td [hunk 3]
```diff
@@ -30,4 +30,106 @@ class XeGPUTypeDef<string name, string typeMnemonic, list<Trait> traits = [],
...
+  let description = [{
+    TensorDesc is a type designed to describe regions of the interested data as well as some 
+    features that are unique to Intel hardware. Different with the builtin tensor type in MLIR, 
+    it essentially only contains the meta data, and doesn't hold the data by itself. It is designed 
+    to mainly support 2d block load/store and DPAS (matrix multiplication instruction) on Intel GPU. 
+    It encodes the following information:
+
+    * shape:  the sizes/shape of the intereted data block, e.g., 8x16 means 8 rows
+              and each row contains 16 continious data element. The rows could be
+              either continuous or not, depends on whether the encoding attribute
```
Comment (adam-smnk) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td:45]: nit: continuous -> contiguous

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 4]
```diff
@@ -23,4 +39,199 @@ class XeGPU_Op<string mnemonic, list<Trait> traits = []>:
           Op<XeGPU_Dialect, mnemonic, traits>;
 
 
+def XeGPU_CreateNdDescOp: XeGPU_Op<"create_nd_tdesc", [Pure, AttrSizedOperandSegments]> {
```
[the remark this answers is not in the document]
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td]: @adam-smnk I added it and `ViewLikeOpInterface` too. It is slightly different from some tensor ops, in that it doesn't have `static_sizes` and `static_strides` arguments, instead they are either from source if it is MemRefType or array of `ShapedType::kDynamic`.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td (hunk 1 above)
[the remark this answers is not in the document]
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td]: @adam-smnk thank for your feedback. See above comments.

Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td]: We also have a set of scattered operators, which are its users. The PR for them is coming after this one. But I updated the verfier based on your feedback. Thanks

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td (hunk 2 above)
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:182]: @adam-smnk Thanks, I updated them

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td (hunk 3 above)
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td]: @adam-smnk Thanks, they are fixed now.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 5]
```diff
@@ -23,4 +26,227 @@ class XeGPU_Op<string mnemonic, list<Trait> traits = []>:
           Op<XeGPU_Dialect, mnemonic, traits>;
 
 
+def XeGPU_CreateNdDescOp: XeGPU_Op<"create_nd_tdesc", [Pure, ViewLikeOpInterface, 
+                        AttrSizedOperandSegments, OffsetSizeAndStrideOpInterface]> {
+
+  let summary = "create nd tensor descriptor operation";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:32]:
```suggestion
  let summary = "Create nd-tensor descriptor operation";
```

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 6]
```diff
@@ -23,4 +26,227 @@ class XeGPU_Op<string mnemonic, list<Trait> traits = []>:
...
+  let arguments = (ins XeGPU_TensorDesc: $TensorDesc,
+                       OptionalAttr<XeGPU_CacheHintAttr>: $l1_hint,
+                       OptionalAttr<XeGPU_CacheHintAttr>: $l2_hint,
+                       OptionalAttr<XeGPU_CacheHintAttr>: $l3_hint);
+
+  // Format: xegpu.prefetch_nd %tdesc {l1_hint = #xegpu.cache_hint<cached>, 
+  //                                   l2_hint = #xegpu.cache_hint<cached>, 
+  //                                   l3_hint = #xegpu.cache_hint<cached>}
+  //         : !xegpu.tensor_desc<8x16xf16>
+  let assemblyFormat = "$TensorDesc attr-dict `:` qualified(type($TensorDesc))";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:192]: Can you always split `prop-dict` out of `attr-dict`? We're trying to deprecate merging the two (it's a slow progress...)

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 7]
```diff
@@ -23,4 +26,227 @@ class XeGPU_Op<string mnemonic, list<Trait> traits = []>:
...
+  let arguments = (ins XeGPU_ValueType: $value,
+                       XeGPU_TensorDesc: $TensorDesc,
+                       OptionalAttr<XeGPU_CacheHintAttr>: $l1_hint,
+                       OptionalAttr<XeGPU_CacheHintAttr>: $l2_hint,
+                       OptionalAttr<XeGPU_CacheHintAttr>: $l3_hint);
+
+  // Format: xegpu.store_nd %3, %2 {l1_hint = #xegpu.cache_hint<uncached>,
+  //                                l2_hint = #xegpu.cache_hint<write_back>, 
+  //                                l3_hint = #xegpu.cache_hint<write_through>}
+  //                                : vector<8x16xf16>, !xegpu.tensor_desc<8x16xf16>
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:247]: This kind of format is better provided as markdown examples in the `let description = ` field (so it shows up on the website as well).
(like you did for the `XeGPU_TensorDesc` type below)

**Review by joker-eph — APPROVED** [review]

**Review by adam-smnk — APPROVED** [review]
Thanks for the tweaks.
Looks pretty good 👍

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td (hunk 6 above)
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:192]: Hi @joker-eph, thanks for the feedback. This is my first time to hear about this, I find some simple examples but didn't find related document about how it works. Do you mind sharing some ideas or docs, if have, about it?

Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:202]: @joker-eph never mind, I got some idea from your RFC https://discourse.llvm.org/t/rfc-introducing-mlir-operation-properties/67846, shared by my colleague

**Comment by joker-eph** [comment]
One nit: please make sure your commit messages are clean before landing, see how it shows up after landing for example here: daebe5c4f27ba140ac8d13abf41e3fe4db72b91a

**Comment by chencha3** [comment]
@joker-eph Ah, I am very sorry about it. I am going to fix it and will be very careful in the future.

**Pull request merged**